src/actors/apache_scanner/apache_scanner.py

```diff
@@ -0,0 +1,137 @@
+    # Sometimes augeas gives us list of dicts and sometimes only dict.
+    # We want to always have the data in the list.
+    if isinstance(tablelist, dict):
+        tablelist = [tablelist]
+
+    for table in tablelist:
+        for key in table:
+            if (myvalue is None and key == mykey) or \
+               (myvalue is not None and key == mykey and myvalue == table[key]):
+                if block is True:
+                    yield table
+                else:
+                    yield key + "=" + table[key]
+        if 'properties' in table:
+            for generator in searchblock(table['properties'], mykey, myvalue, block):
+                yield generator
+
+
+def getblock(tablelist, mykey, myvalue=None, block=True, propitem=0):
+    """ Accumulates matched data and returns them for output.
+
+        Arguments:
+            block: if False, will pass block=False to searchblock
+            propitem: specify from which index in 'properties' to search
+
+        Returns:
+            Extracted data in form of dictionary.
+    """
+    out = {}
+    if block is True:
+        for dic in searchblock(tablelist, mykey, myvalue, block):
+            if dic[mykey] not in out:
+                out[dic[mykey]] = [dic['properties'][propitem]['value']]
+            else:
+                out[dic[mykey]].append(dic['properties'][propitem]['value'])
+    else:
+        for line in searchblock(tablelist, mykey, myvalue, block):
+            key, value = line.split("=")
+            if key not in out:
+                out[key] = [value]
+            else:
+                out[key].append(value)
+    return out
+
+
+def printall(tablelist):
+    """ Prints all occurrences of the specified key-value pairs.
+
+    Tuple @searches contains information about what to extract and how it will
+    be extracted (see @block and propitem@ arguments to getblock() function.
+
+    Args:
+        tablelist: 'aug_httpd' lens data from the augeas actor output
+
+    Returns:
+        json formatted extracted data from the augeas aug_httpd lens
+    """
+    searches = (
+        # Arguments for the getblock function.
+        # mykey, myvalue=None, block=True, propitem=0
+        ("value", "IncludeOptional"),
+        ("value", "User"),
+        ("value", "Group"),
+        ("value", "Mutex"),
+        ("value", "PidFile"),
+        ("value", "Listen"),
+        ("value", "TypesConfig"),
+        ("value", "CacheRoot"),
+        ("value", "MIMEMagicFile"),
+        ("value", "ScriptSock"),
+        ("value", "SSLCertificateKeyFile"),
+        ("value", "SSLCertificateFile"),
+        ("value", "CustomLog"),
+        ("value", "ErrorLog"),
+        ("value", "Include"),
+        ("name", "VirtualHost"),
+        ("name", "Directory"),
+        ("value", "DocumentRoot"),
+        ("absolute_path", None, False),
+        ("value", "LoadModule", True, 1),
+    )
+
+    out = {}
+    for args in searches:
+        result_table = getblock(tablelist, *args)
```

Comment:
While this is cool and (IMO) `getblock` and `searchblock` are well-coded, it was a little bit difficult for me to follow through the code.

except by returning two types of data

For example, looking at the code as is, I don't understand why sometimes you don't want to return whole dicts in `searchblock()` (`block` parameter). I'm sure you have a reason for that, however, if you could get rid of the block handling the scanner actor will be simpler.

If it's not possible to get rid of that, and since I don't have a better suggestion, I'd be OK with leaving it as is.

block: if False, do not yield whole dict but only key-value pair

Reply:
This is true. I will try to refactor this thing a little bit.